# Data Wrangling


- Data wrangling also known as data munging, is the process of transforming and mapping of data from its atomic data form into a format whereby it makes the extracted information more apppropriate and valuable for a variety of purposes such as analytics.

## Wrangling Data by:

- Merging data frames
- Filtering data frames
- Manipulating rows and columns

## Importing essential python libraries 


In [3]:
import pandas as pd

## Creating a data frame from importing the personality score dataset

In [4]:
pscore = pd.read_csv('/home/rulani/Downloads/Data-wrangling-master/personality_scores.csv',delimiter=";")
pscore.dropna(axis='columns').head()

ID Section 5 of 6 [I am always prepared.]  \
0   0                                 (3, 5)   
1   1                                 (3, 5)   
2   2                                 (3, 5)   
3   3                                 (3, 5)   
4   4                                 (3, 3)   

  Section 5 of 6 [I am easily disturbed.]  \
0                                  (4, 5)   
1                                  (4, 5)   
2                                  (4, 3)   
3                                  (4, 5)   
4                                  (4, 5)   

  Section 5 of 6 [I am exacting (demanding) in my work.]  \
0                                             (3, 5)       
1                                             (3, 5)       
2                                             (3, 3)       
3                                             (3, 3)       
4                                             (3, 3)       

  Section 5 of 6 [I am full of ideas.]  \
0                               (5, 5)   
1                               (5, 5)   
2                               (5, 5)   
3                               (5, 5)   
4                               (5, 3)   

  Section 5 of 6 [I am interested in people.]  \
0                                      (2, 3)   
1                                      (2, 5)   
2                                      (2, 5)   
3                                      (2, 5)   
4                                      (2, 3)   

  Section 5 of 6 [I am not interested in abstract ideas.]  \
0                                             (5, 3)        
1                                             (5, 3)        
2                                             (5, 5)        
3                                             (5, 3)        
4                                             (5, 3)        

  Section 5 of 6 [I am not interested in other people's problems.]  \
0                                             (2, 3)                 
1                                             (2, 5)                 
2                                             (2, 5)                 
3                                             (2, 3)                 
4                                             (2, 3)                 

  Section 5 of 6 [I am not really interested in others.]  \
0                                             (2, 5)       
1                                             (2, 5)       
2                                             (2, 5)       
3                                             (2, 3)       
4                                             (2, 3)       

  Section 5 of 6 [I am quick to understand things.]  ...  \
0                                            (5, 5)  ...   
1                                            (5, 5)  ...   
2                                            (5, 5)  ...   
3                                            (5, 3)  ...   
4                                            (5, 5)  ...   

  Section 5 of 6 [I often forget to put things back in their proper place]  \
0                                             (3, 5)                         
1                                             (3, 5)                         
2                                             (3, 5)                         
3                                             (3, 1)                         
4                                             (3, 5)                         

  Section 5 of 6 [I pay attention to details.]  \
0                                       (3, 5)   
1                                       (3, 1)   
2                                       (3, 5)   
3                                       (3, 5)   
4                                       (3, 5)   

  Section 5 of 6 [I seldom feel blue (down).]  \
0                                      (4, 3)   
1                                      (4, 1)   
2                                      (4, 1)   
3                                      (4, 1)   
4        

## Examining data for :

- Uniqueness of ID 
- Availability of NaN entires

In [8]:
# dropping ALL duplicte values 
dropped_pscore = pscore.drop_duplicates(subset ="ID", 
                     keep = False, inplace = False) 

In [10]:
#Checking the lengths of the datasets using assert

def same():
    assert len(pscore) == len(dropped_pscore)
    return True
       
same()

True

In [11]:
print('There are no duplicated IDs in this data frame since '
    '\nThe length of dropped duplicate data frame is,' ,len(dropped_pscore),
    '\nand that of the original data frame is', len(pscore), '.')

There are no duplicated IDs in this data frame since 
The length of dropped duplicate data frame is, 1555 
and that of the original data frame is 1555 .


In [7]:
# making new data frame with dropped NA values 
new_dropped_pscore = dropped_pscore.dropna(axis = 1, how ='any', inplace = False)

new_dropped_pscore.head()

ID Section 5 of 6 [I am always prepared.]  \
0   0                                 (3, 5)   
1   1                                 (3, 5)   
2   2                                 (3, 5)   
3   3                                 (3, 5)   
4   4                                 (3, 3)   

  Section 5 of 6 [I am easily disturbed.]  \
0                                  (4, 5)   
1                                  (4, 5)   
2                                  (4, 3)   
3                                  (4, 5)   
4                                  (4, 5)   

  Section 5 of 6 [I am exacting (demanding) in my work.]  \
0                                             (3, 5)       
1                                             (3, 5)       
2                                             (3, 3)       
3                                             (3, 3)       
4                                             (3, 3)       

  Section 5 of 6 [I am full of ideas.]  \
0                               (5, 5)   
1                               (5, 5)   
2                               (5, 5)   
3                               (5, 5)   
4                               (5, 3)   

  Section 5 of 6 [I am interested in people.]  \
0                                      (2, 3)   
1                                      (2, 5)   
2                                      (2, 5)   
3                                      (2, 5)   
4                                      (2, 3)   

  Section 5 of 6 [I am not interested in abstract ideas.]  \
0                                             (5, 3)        
1                                             (5, 3)        
2                                             (5, 5)        
3                                             (5, 3)        
4                                             (5, 3)        

  Section 5 of 6 [I am not interested in other people's problems.]  \
0                                             (2, 3)                 
1                                             (2, 5)                 
2                                             (2, 5)                 
3                                             (2, 3)                 
4                                             (2, 3)                 

  Section 5 of 6 [I am not really interested in others.]  \
0                                             (2, 5)       
1                                             (2, 5)       
2                                             (2, 5)       
3                                             (2, 3)       
4                                             (2, 3)       

  Section 5 of 6 [I am quick to understand things.]  ...  \
0                                            (5, 5)  ...   
1                                            (5, 5)  ...   
2                                            (5, 5)  ...   
3                                            (5, 3)  ...   
4                                            (5, 5)  ...   

  Section 5 of 6 [I often forget to put things back in their proper place]  \
0                                             (3, 5)                         
1                                             (3, 5)                         
2                                             (3, 5)                         
3                                             (3, 1)                         
4                                             (3, 5)                         

  Section 5 of 6 [I pay attention to details.]  \
0                                       (3, 5)   
1                                       (3, 1)   
2                                       (3, 5)   
3                                       (3, 5)   
4                                       (3, 5)   

  Section 5 of 6 [I seldom feel blue (down).]  \
0                                      (4, 3)   
1                                      (4, 1)   
2                                      (4, 1)   
3                                      (4, 1)   
4        

## Creating a function for the total score of each of the personality test subscales

In [165]:
def assigning(row,Category_Num):
    count = 0
    asd = 0
    for column in new_dropped_pscore.columns:
        tuplee = row[column]
        if type(tuplee) == str:
            catergory = tuplee[1]
            result = int(tuplee[4])
            if catergory == Category_Num:
                count += result
    return count

## Creating a column for the Big-Five Factor Markers

In [175]:
new_dropped_pscore['Extraversion'] = dropped_pscore.apply(lambda row : assigning(row,'1'), axis = 1)
new_dropped_pscore['Agreeableness'] = dropped_pscore.apply(lambda row : assigning(row,'2'), axis = 1)
new_dropped_pscore['Conscientiousness'] = dropped_pscore.apply(lambda row : assigning(row,'3'), axis = 1)
new_dropped_pscore['Emotional Stability'] = dropped_pscore.apply(lambda row : assigning(row,'4'), axis = 1)
new_dropped_pscore['Openness to experiences'] = dropped_pscore.apply(lambda row : assigning(row,'5'), axis = 1)

In [242]:
#Displaying the data frame incoporated with the big five factor Markers

new_dropped_pscore[['ID','Extraversion','Agreeableness',
                    'Conscientiousness','Emotional Stability',
                    'Openness to experiences']].tail()

ID  Extraversion  Agreeableness  Conscientiousness  \
1550  1550            32             48                 38   
1551  1551            40             48                 40   
1552  1552            42             44                 40   
1553  1553            36             38                 48   
1554  1554            22             44                 42   

      Emotional Stability  Openness to experiences  
1550                   44                       44  
1551                   44                       34  
1552                   46                       48  
1553                   36                       44  
1554                   40                       30

## Creating the data frame from importing the department dataset

In [8]:
department = pd.read_csv('/home/rulani/Downloads/Data-wrangling-master/departments .csv',delimiter=";")
department.head()

ID Department
0   0       Data
1   1       Data
2   2       Data
3   3       Data
4   4       Data

## Examining data for :

- The shape of the data frame 
- Availability of NaN entires

In [184]:
#Checking the shape of the data

department.shape

(1555, 2)

- This illustrates that there are 1555 rows and 2 columns 

In [185]:
#Checking for nan values in dataset

department.isnull().sum()

ID            0
Department    0
dtype: int64

- The output above shows that there are no nan values within our dataset

## Merging Department dataframe with the personality score dataframe

In [208]:
new_combined_data = pd.merge(left=dropped_pscore, right=department, left_on='ID', right_on='ID')

new_combined_data[['ID','Extraversion','Agreeableness',
                    'Conscientiousness','Emotional Stability',
                    'Openness to experiences','Department']].tail()

ID  Extraversion  Agreeableness  Conscientiousness  \
1550  1550            32             48                 38   
1551  1551            40             48                 40   
1552  1552            42             44                 40   
1553  1553            36             38                 48   
1554  1554            22             44                 42   

      Emotional Stability  Openness to experiences Department  
1550                   44                       44    Web dev  
1551                   44                       34    Web dev  
1552                   46                       48    Web dev  
1553                   36                       44    Web dev  
1554                   40                       30    Web dev

- The table above shows the last 5 entries in the merged data frame

## Asserting for the equality of the lengths of the department and recently merged data frame

In [192]:
def sameLength(length):
    assert len(department) == len(new_combined_data)
    return True
       
sameLength('')

True

- This shows that the length of the department and the recently merged data frame(named new_combined_data) are equall , implying that there are no additional entires made

## Applicants who scored less than 30 on emotional stability, conscientiousness and agreeableness test.

In [255]:
# Computations for appicants' score

subset = new_combined_data[
    (new_combined_data['Emotional Stability'] < 30) & 
    (new_combined_data['Conscientiousness'] < 30) &
    (new_combined_data['Agreeableness'] < 30)]


subset[['ID','Agreeableness','Conscientiousness',
        'Emotional Stability','Department']]


ID  Agreeableness  Conscientiousness  Emotional Stability   Department
881    881             28                 26                   28         Data
1197  1197             22                 26                   26  Copywriting

- The table above illustrates that applicant number(ID) 881 in the Data department
and applicant number 1197 in the Copywriting department scored less than 30 for emotional stability, conscientiousness including agreeableness.

## Creating a Risk Level column for each of the observations

In [206]:
#Assigning Risk Level to the new_combined_data, data frame

low_risk_IDs = []
high_risk_IDs = []
new_combined_data['Risk Level'] = "Low Risk"
new_combined_data_length = len(combined_data)


for id in range(combined_data_length):

    if combined_data['Conscientiousness'][id] < 30 and new_combined_data.Agreeableness[id] < 30 and combined_data['Emotional Stability'][id] < 30:
         high_risk_IDs.append(id)
         combined_data['Risk Level'][id] = "High Risk"
    if combined_data['Conscientiousness'][id] > 30 and new_combined_data.Agreeableness[id] > 30 and combined_data['Emotional Stability'][id] > 30:
         low_risk_IDs.append(id)  
         combined_data['Risk Level'][id].replace("no risk", "Low Risk")

new_combined_data[['ID','Department','Risk Level']].head()


/home/rulani/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:10: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  # Remove the CWD from sys.path while we load stuff.


ID Department Risk Level
0   0       Data   Low Risk
1   1       Data   Low Risk
2   2       Data   Low Risk
3   3       Data   Low Risk
4   4       Data   Low Risk

## Brief description of the Risk Level in the observation

In [213]:
counting = new_combined_data['Department'].groupby(combined_data['Risk Level'])

counting.size()

Risk Level
High Risk       2
Low Risk     1553
Name: Department, dtype: int64

- This shows that in the entire data frame, 2 entries are high risk whilst 1553 are Low risk

## Risk Levels as per Department

In [225]:
total = combined_data.groupby(['Risk Level', 'Department']).count()
total.reset_index(inplace = True)
total[['ID','Risk Level','Department']]

ID Risk Level   Department
0    1  High Risk  Copywriting
1    1  High Risk         Data
2  325   Low Risk  Copywriting
3  328   Low Risk         Data
4  120   Low Risk       Design
5  449   Low Risk     Strategy
6  170   Low Risk      Web Dev
7  161   Low Risk      Web dev

- The table above shows the amount of high risk applicants and low risk applicants in each department

## Detailed data frame for risk levels as per department

In [234]:
#Creating a pivot table to show risk levels as per department

table = pd.pivot_table(combined_data,index=['Risk Level'],columns=['Department'],aggfunc='count').fillna(0.0).astype(int)

table[['Agreeableness', 'Conscientiousness', 'Emotional Stability']]

Agreeableness                                       \
Department   Copywriting Data Design Strategy Web Dev Web dev   
Risk Level                                                      
High Risk              1    1      0        0       0       0   
Low Risk             325  328    120      449     170     161   

           Conscientiousness                                       \
Department       Copywriting Data Design Strategy Web Dev Web dev   
Risk Level                                                          
High Risk                  1    1      0        0       0       0   
Low Risk                 325  328    120      449     170     161   

           Emotional Stability                                       
Department         Copywriting Data Design Strategy Web Dev Web dev  
Risk Level                                                           
High Risk                    1    1      0        0       0       0  
Low Risk                   325  328    120      449     170     161